### DISCUSSION TITLE
Android test at scale - open source projects

### DISCUSSION THREAD

**Discussion by bootstraponline** [MobileNativeFoundation/discussions#85] [General]
What excellent open source projects are there for Android testing at scale? It'd be great to highlight all the current projects in this space.

The work @Malinskiy is doing is awesome!

- https://github.com/Malinskiy/marathon
  - https://malinskiy.medium.com/adam-a-second-birth-to-androids-ddmlib-c90fdde4c39d
  - https://malinskiy.medium.com/im-not-going-to-use-ktor-as-a-network-client-anymore-here-s-why-3b2560c590b7

The Flank team is making progress on AWS support. 🙂 
- https://github.com/Flank/flank

Links
- https://github.com/Banno/Gordon
- https://github.com/AdevintaSpain/Barista
- https://github.com/linkedin/test-butler
- https://github.com/avito-tech/avito-android
- https://github.com/google/android-emulator-container-scripts

**Comment by Malinskiy** [comment]
Testing at scale for any tech stack involves a lot of moving pieces. When I tackled this problem in agoda, I've worked at all the levels from hardware to the test tooling and test frameworks.

Some of these problems can be outsourced or replaced with cloud service, provided you accept their limitations and trade-offs.

A typical solution consists of the following:

### Compute
This is the backbone of running the whole testing at scale: the assumption being "if you buy 50% more hardware - you get 50% faster testing". It is quite standard now to use elastic CI runners and they usually should be enough for running tests that do not need Android environment (real device/emulator). In practice, even this piece requires attention as your test count reaches 10k or even 100k. Achieving success here requires careful planning on the hardware: can we scale tests horizontally or is there a limit? Is it better to scale vertically (e.g. AWS z1d instances)?
Your compute should also provide some form of Android environment: real devices + emulators. Without going into the specifics of each app development requirements, your testing infrastructure ideally should support both real and emulated environments. [Devicefarmer](https://github.com/DeviceFarmer/stf) (ex OpenSTF) is an open-source project that provides a unified API to access Android compute. Deployment of such a system is a complex task which I tried to optimize in a now-abandoned [android-farm](https://github.com/agoda-com/android-farm) that installed the whole farm via helm in one command on kube cluster. Other solutions to the same problem involve using proprietary device farms (Firebase Test Lab, Amazon Device Farm, Anbox Cloud, etc). Each of them has tradeoffs: do you have direct access to Android OS? What's the cost? Each of those questions needs to be evaluated as the development team grows and new test cases are added.
### Test frameworks
Writing test code at scale is a hard problem. Getting rid of false-positive and false-negatives in testing is a challenge in part because we don't share stable code in tests. This can be mitigated to some extent by using design patterns, but it's hard to synchronize design patterns across teams without slowing things down. For UI tests one of the solutions is using page object pattern which allows sharing a stable-ish code (e.g. press a button) between all the tests. This is not a problem if you have dozens of tests, but it becomes extremely important as you reach your first thousand. To sync the page object patterns at agoda we wrote [Kakao](https://github.com/agoda-com/Kakao) which later was built upon in [Kaspresso](https://github.com/KasperskyLab/Kaspresso) with the idea that each particular action can be flaky and needs to be repeated. Although it masks problems in tests and can be detrimental if applied in the wrong way, it allows teams to still continue working at scale and have a time window to fix their tests while the CI environment continues to function.
### Test runner
After scaling the compute and figuring out the frameworks part, you're still left with a problem that the assumption of "add 50% more hardware -> tests become 50% faster" doesn't hold true. Some of the test runners depend on direct ADB access that allows you to have control over the test execution flow ([spoon](https://github.com/square/spoon), [fork](https://github.com/shazam/fork), [composer](https://github.com/gojuno/composer), [avito-runner](https://github.com/avito-tech/avito-android)), while others do not have that luxury and wrap whatever their respective compute allows them to do like [flank](https://github.com/Flank/flank).

The biggest problem with custom runners is that all of them need to have a dependency on compute in some shape or form and deploying + supporting 100 real devices with a 100 emulators is a challenge. That's why a lot of teams are limited by being able to run emulators on CI agents and hence can't grow past a certain level.

On top of the above, there are also a lot of problems with supporting: you need metrics, alerts, flakiness analysis, but those deserve a separate discussion.

For those interested, here is a [general overview](https://www.youtube.com/watch?v=JrxOk20tf9k) of the problem space and how it was tackled in agoda.
